builtins: handle builtins involving string casts to geometry

otan opened the pull request:
Release note (sql change): Implement geospatial functions that take in a
string argument such that when an ambiguous function that could be
either geometry or geography is encountered, geometry is chosen. This is
the case for the following functions:
* st_area
* st_asewkt
* st_asgeojson
* st_buffer
* st_coveredby
* st_covers
* st_distance
* st_dwithin
* st_intersects
* st_length

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51563)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-:-MCgG0xO5KPkymZ4gjCG:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 3408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgBo7b18VqNU50TFnG:-MCgBo7b18VqNU50TFnH:bxep7ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/logictest/testdata/logic_test/geospatial#L3408)):*
> ```
> 0  LINESTRING (0 0, 1 1, 2 2)           {"type":"LineString","coordinates":[[0,0],[1,1],[2,2]]}             LINESTRING (0 0, 1 1, 2 2)           2.82842712474619
> 
> # Test for string against with 2+ arguments cast to GEOMETRY.
> ```

string against with? not sure what this is saying...
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3467 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgCR6dCN_nejs3eaPN:-MCgCR6dCN_nejs3eaPO:bjppcpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3467)):*
> ```Go
> 		"st_intersects",
> 		// "st_intersection",
> 		"st_length",
> ```

Maybe add a TODO above this for the commented out functions? Also would help to have a comment explaining the purpose of this block.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgFfDq4_r-3ItOXaGA:-MCgFfDq4_r-3ItOXaGB:b-45v6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3580)):*
> ```Go
> // appendStrArgOverloadForGeometryArgOverloads appends overloads by casting string
> // types into geometry types if any geometry arguments are taken in as input.
> // This allows PostGIS to "prioritize" the string overloads when resolving
> ```

How does this prioritization work? A bit more detail here would help.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgDnJSC9BjY-TCh8pr:-MCgDnJSC9BjY-TCh8ps:bwmtcky) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3628)):*
> ```Go
> 				arg := string(tree.MustBeDString(args[i]))
> 				var g *geo.Geometry
> 				g, err = geo.ParseGeometry(arg)
> ```

shouldn't you check the error here?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 3408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgBo7b18VqNU50TFnG:-MCgr8QQ2mLbvsfwWbde:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/logictest/testdata/logic_test/geospatial#L3408)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

string against with? not sure what this is saying...
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3467 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgCR6dCN_nejs3eaPN:-MCgt8GE3NLZ3PMvC-Zb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3467)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe add a TODO above this for the commented out functions? Also would help to have a comment explaining the purpose of this block.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgFfDq4_r-3ItOXaGA:-MCgrbXgBT-BznBxlo4W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3580)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How does this prioritization work? A bit more detail here would help.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgDnJSC9BjY-TCh8pr:-MCgrI868CgDRsPhAiaq:brfn8xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3628)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

shouldn't you check the error here?
</blockquote></details>

it gets checked after the closure.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 3 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgDnJSC9BjY-TCh8pr:-MChBbyK9itgJvWktH0L:b-2a58o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3628)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

it gets checked after the closure.
</blockquote></details>

But not before you pass `g` to `NewDGeometry` below... why not just move the `if` statement from line 3623 here?


<!-- Sent from Reviewable.io -->

otan committed:
builtins: handle builtins involving string casts to geometry

Release note (sql change): Implement geospatial functions that take in a
string argument such that when an ambiguous function that could be
either geometry or geography is encountered, geometry is chosen. This is
the case for the following functions:
* st_area
* st_asewkt
* st_asgeojson
* st_buffer
* st_coveredby
* st_covers
* st_distance
* st_dwithin
* st_intersects
* st_length

otan commented:
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-MCgDnJSC9BjY-TCh8pr:-MChGd9w5BwRUhXZJWDr:bku2iov) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8cc4557ebd4d75fcc38d9fb7bfee8749cdbcc0/pkg/sql/sem/builtins/geo_builtins.go#L3628)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

But not before you pass `g` to `NewDGeometry` below... why not just move the `if` statement from line 3623 here?
</blockquote></details>

lol, woops!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51563#-:-MChI6rx9fwJMMElMaLp:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
bors r=rytaft
